tests/python/mkl/test_mkldnn.py

```diff
@@ -89,6 +91,33 @@ def get_tensors(args, shapes, ctx):
     except:  # pylint: disable=bare-except
         assert 0, "test_mkldnn_model exception in bind and execution"
 
+def test_mkldnn_engine_threading():
+    """
+    This test will trigger mkldnn engine on different thread of execution
+    """
+
+    import mxnet as mx
+    from mxnet import gluon, nd
+
+    net = gluon.nn.HybridSequential()
+    with net.name_scope():
+        net.add(gluon.nn.Conv2D(channels=32, kernel_size=3, activation=None))
+    net.collect_params().initialize(ctx=mx.cpu())
+
+    val_data = gluon.data.DataLoader(
+        gluon.data.vision.CIFAR10(train=False),
+        batch_size=32, shuffle=False, num_workers=1)
```

Comment:
can you remove this data loader? it doesn't seem this test needs to use it

Reply:
actually, the dataloader is the one that triggers different thread context, i tried without a dataloader (or even if i pass num_workers = 0 to above dataloader) it runs on same thread, so bug wont happen. Not sure if there is any other way to trigger different thread execution context, but this unittest produces the bug currently.

Reply:
Thanks a lot for this test! Could you please elaborate the exact behaviour of this unit test in the test with a block comment. I agree with Da that at the moment, it's hard to grasp the exact problem from reading the code. For me it's hard to understand when different threads are getting started and what the exact issue is.

Reply:
yes, please provide comments why we need data loader here. Thanks

Reply:
Gluon DataLoader allows us to create a new thread as it iterates over data batch. Added comments to PR.